### ISSUE TITLE
encoding/xml: namespaced and non-namespaced attributes conflict

### ISSUE THREAD

**Issue by ainar-g** [golang/go#11724]
Go versions:

```
go version devel +2e4b659 Wed Jul 15 06:04:51 2015 +0000 linux/amd64

go version go1.4.2 linux/amd64
```

Code: http://play.golang.org/p/S2LU2vfk4i.

Expected: no error and `main.Discount{XSIType:"ProgressivePromotion", Type:"Percent", From:"2015-07-28T00:00:00", To:"2015-07-30T00:00:00"}`

Got: `error: main.Discount field "XSIType" with tag "xsi type,attr" conflicts with field "Type" with tag "type,attr"`

Found by Eugene Toropov [at golang-ru](https://groups.google.com/d/msg/golang-ru/qoNDWMFR988/ellUIv9X3CEJ). May be related to #3703.

**Comment by rogpeppe** [comment]
This is unfortunate. The existing Go convention is that if no namespace is specified
in the struct tag, it will match any namespace including the empty namespace.

Changing that will break existing code.

One possibility might be to make the tag `xml:" name,attr"` distinct from `xml:"name,attr"`
but that seems way too subtle.
Another possibility might be to use a special namespace to signify the empty namespace.
Various portions of the namespace are reserved (for example the xml-prefixed namespace),
so something like:

   `xml:"xmlempty name,attr"`

might be possible, though ugly.

**Comment by ainar-g** [comment]
>The existing Go convention is that if no namespace is specified
in the struct tag, it will match any namespace including the empty namespace.

Is this documented anywhere? I might've missed that, but [the package docs](http://golang.org/pkg/encoding/xml/) don't seem to mention that. Might be a documentation issue.

Also, why can't that be changed to "if no namespace is specified in the struct tag, it will match any namespace including the empty namespace *but excluding fields with the same name that explicitly specify their namespace*" (or something better worded)? This doesn't seem like it would break anyone's code, because right now having `xml:"foo,attr"` and `xml:"bar foo,attr"` in the same struct results in an error.

**Comment by rogpeppe** [comment]
> Is this documented anywhere? I might've missed that, but the package docs don't seem to mention that. Might be a documentation issue.

Yes - this could definitely do with better docs - please feel free to raise an issue.

> Also, why can't that be changed to "if no namespace is specified in the struct tag, it will match any
> namespace including the empty namespace *but excluding fields with the same name that
> explicitly specify their namespace*" (or something better worded)? This doesn't seem like it would
> break anyone's code, because right now having `xml:"foo,attr"` and `xml:"bar foo,attr"` in the same
> struct results in an error.

I'm not sure that would work very well. If I'm trying to unmarshal from an attribute
in the empty namespace, I don't want any interference from other attributes that
I never heard of before. For example in the code linked above, if the code just
wants the `type` attribute and not any others such as `xsi:type` then there's
no way to do that currently.

This is a particular problem because attributes are in the empty namespace by
default, even if there's a default namespace set up (XML is weird).

**Comment by ainar-g** [comment]
>Yes - this could definitely do with better docs - please feel free to raise an issue.

Raised  #11735.

If the convention won't be changed (at least, not until Go 2.0), then we could use the form `xml:"- name"` to explicitly specify an empty namespace. This is more explicit than a single space (which might as well break existing code), looks consistent with `xml:"-"`, and doesn't collide with anything since XML names [can't start with "-"](http://www.w3.org/TR/REC-xml/#NT-Name).

Still, when there's a change to change the convention (e.g. 2.0), I'd suggest to make `xml:"name"` match only empty namespace, while something like `xml:"* name"` could be a wildcard to match any namespace.

**Comment by mmcdole** [comment]
@rogpeppe just ran into this issue myself trying to unmarshal RSS documents which use the "itunes" namespace.  My documents will sometimes have ```<itunes:author>``` and ```<author>``` elements.  

Until Go's behavior is changed, what can we do to work around this?

**Comment by ainar-g** [comment]
@mmcdole Such questions are better off on [StackOverflow](http://stackoverflow.com/questions/tagged/go), golang-nuts, or other forums. Also more people would be able to help you if you provide an [MCVE](http://stackoverflow.com/help/mcve).

The workaround I originally proposed at [golang-ru](https://groups.google.com/d/msg/golang-ru/qoNDWMFR988/4vs7pW2OZNgJ) is to replace all namespaced attributes in test with a modified one, i.e. replace `<itunes:author>` with `<itunes:_author>` and then unmarshal it into a separate field. I don't know whether it could help you.

**Comment by pdw-mb** [comment]
This looks like a duplicate of #8535 .

**Comment by iWdGo** [comment]
In the first example, the namespace is not bound. This is mandatory (https://www.w3.org/TR/xml-names/#dt-NSName). If fixed, this issue is a duplicated of #8535 and the same fix submitted applies.

**Comment by iwdgo** [comment]
The submitted playground example now succeeds. Issue was incidentally fixed by [CL 106575](https://go-review.googlesource.com/c/go/+/106575) for #8535 in Go1.21. Issue can probably be closed.

**Comment by ianlancetaylor** [comment]
Thanks.
